### ISSUE TITLE
Update to newer kustomize 

### ISSUE THREAD

**Issue by dims** [kubernetes/kubernetes#82905]
v3.2.0 of kustomize was released yesterday, we are still at v2.0.3+ Can we please update?
https://github.com/kubernetes-sigs/kustomize/releases

(Will also  help with https://github.com/kubernetes/kubernetes/issues/77024)

/sig cli
/sig cluster-lifecycle

**Comment by liggitt** [comment]
cc @Liujingfang1

**Comment by Liujingfang1** [comment]
I've considered this for a while. Since the dependency issue, we couldn't re-integrate kustomize as it is into kubectl. Need to split Kusotmize into multiple Go modules first. Will write a proposal for this.

**Comment by rosscdh** [comment]
Seems the module split took place? Be grand to bring this up to date.

Great work on kustomize by the way it is THE way to work with kube

**Comment by Liujingfang1** [comment]
The re-factorization of Kustomize is in progress. This will make the integration into kubectl easier. cc @monopole

**Comment by dims** [comment]
@Liujingfang1 can we please pick this back up again?

**Comment by munnerz** [comment]
I just started to take a look into how easy this is (as we're about to begin depending on `k8s.io/kubectl` in our own project which will then pull this in too..)

First blocker I ran into is https://github.com/kubernetes-sigs/kustomize/blob/v3.3.1/kustomize/internal/commands/build/build.go now being part of an `internal` package, whereas previously `sigs.k8s.io/kustomize/pkg/commands/build` could be imported. This bit of code is needed by `staging/src/k8s.io/cli-runtime/pkg/kustomize` in builder.go:

```
// RunKustomizeBuild runs kustomize build given a filesystem and a path
func RunKustomizeBuild(out io.Writer, fSys fs.FileSystem, path string) error {
	f := k8sdeps.NewFactory()
	o := build.NewOptions(path, "")
	return o.RunBuild(out, fSys, f.ResmapF, f.TransformerF)
}
```

**Comment by dims** [comment]
@seans3 @pwittrock @monopole @soltysh ping! i started an email thread and pinged you all on slack as well.

**Comment by pwittrock** [comment]
@dims Thanks for the ping.  Quick update on the status for this.

TLDR: we are actively working towards this on 2 fronts.

## Background

IIUC from @monopole, the migration to go-modules made updating kustomize as it exists today infeasible.

We've explored a number of approaches updating kustomize while keeping the dependencies on apimachinery over the past 9 months, none of which has panned out.

## Status

To update kustomize in kubectl we must do either of

1) Complete the migration of kubectl out of staging and into the kubectl repo.  The code for this is nearly complete.  However we expect to encounter unknown-unknowns in the process of fully moving it out.  This probably requires coordination with sig-release + sig-testing since it will impact both.

2) Strip out all dependencies on kubernetes/staging dependencies from kustomize.  This is actively being pursued -- rebasing kustomize on its own machinery libraries for working with yaml configuration.  

- https://github.com/kubernetes-sigs/kustomize/issues/2340
- https://github.com/kubernetes-sigs/kustomize/issues/2356

We are actively working on both these, but neither is trivial.

**Comment by larstiq** [comment]
How is this going? Today I ran into `kustomize` v3.5.4 wanting `git::ssh://user@host:port/group/repo.git//path/` to specify a remote base reachable via a non-standard ssh port, but `kubectl` isn't happy with it. I have yet to find a way to write that so that both `kustomize` and `kubectl` will do the right thing.

**Comment by tylerauerbeck** [comment]
Has there been any progress on this?

**Comment by soltysh** [comment]
It's still a thing sig-cli is working on.
/lifecycle frozen

**Comment by neolit123** [comment]
we are planning to remove the kustomize imports from kubeadm in 1.20. there is an alternative (alpha) similar to the patch types that `kubectl patch` supports.
/remove-sig cluster-lifecycle

**Comment by tdensmore** [comment]
Given the amount of time that issues concerning `kustomize` versions have been open ([here](https://github.com/kubernetes-sigs/kustomize/issues/1424https://github.com/kubernetes-sigs/kustomize/issues/1424) and [here](https://github.com/kubernetes-sigs/kustomize/issues/1500)), would it be easier to integrate _krew_ with `kubectl` by default and make `kustomize` a _krew_ plugin (versioned binary) that is included with `kubectl` by default? 

The plugin could be freely upgraded / downgraded as necessary to allow backward compatibility where needed to prevent breaking CI/CD systems (it might also easily allow the `kustomize -version` command to work out of the box). The default shipping version could be 'pinned' to generate a warning when a non-shipped version of `kustomize` is used (or the kustomize plugin is removed completely), but otherwise would provide the freedom and flexibility to upgrade / downgrade the version as needed ... 

I am **SURE** there is more complexity to this issue that I am aware of, but `kustomize` and `kubectl` seem very tightly coupled ATM which from all outward appearances seems untenable.

**Comment by soltysh** [comment]
We're tracking this in https://github.com/kubernetes/kubectl/issues/818. 
/close
